### PR TITLE
[DOCS]: Add some philosophy.

### LINKED ISSUE #51: [DOCS]: Philosophy of j5

### PR DESCRIPTION
**Pull request by trickeydan**
Closes #51 

When reviewing long prose text like this, it is probably easier in the majority of cases to use the GitHub suggest tool.

I doubt this is perfect. I've never been too great at this sort of writing, but I think it is important to explain the reasons behind the project.

### CODE REVIEW

File: docs/concepts/philosophy.rst [hunk 1]
```diff
@@ -1,2 +1,40 @@
 Philosophy Behind j5
-====================
\ No newline at end of file
+====================
+
+Some Background
+---------------
+
+Student Robotics is a charity that was originally founded by a group of students at the University of Southampton with
```
Comment (mildlyincompetent) [inline docs/concepts/philosophy.rst]: Dodgy new lines cause kidney stones in mail clients, please avoid.

File: docs/concepts/philosophy.rst [hunk 2]
```diff
@@ -1,2 +1,40 @@
...
+the goal of bringing the excitement of engineering and the challenge of coding to young people through robotics. This
+has involved running an annual robotics competition almost every year since 2008, where groups of sixth form students
+are given some robotics kit, time and mentoring to develop a competitive robot for a unique challenge. In order to
+reduce the barrier to entry for the competition, it is essential that a knowledge of low level programming and
+hardware is not required by the students. Thus, a Python API is usually supplied alongside hardware that is developed
+in order to make things easier.
```
Comment (mildlyincompetent) [inline docs/concepts/philosophy.rst]: This implies that Python is somehow easy to understand or reason about. As Python is an imperative language with side effects, this is demonstrably false.

File: docs/concepts/philosophy.rst [hunk 3]
```diff
@@ -1,2 +1,40 @@
...
+
+In 2017 / 18, Student Robotics underwent some restructuring and as a result did not hold a competition. To meet the demand of
+teachers for the competition, volunteers created two independent competitions SourceBots and Robocon. Both competitions
+designed and built their own robotics kits that were very similar to the current Student Robotics kit, yet completely
+incompatible with both each other and the previous kit.
+
+Unification
+-----------
+
+Following the reappearance of Student Robotics to the scene in late 2018, there were now three separate, very similar,
```
Comment (mildlyincompetent) [inline docs/concepts/philosophy.rst]: what scene???

File: docs/concepts/philosophy.rst [hunk 4]
```diff
@@ -1,2 +1,40 @@
...
+and also incompatible robotics kits that were being used for the same purpose. None of the kits were perfect, and volunteers
+didn't want to replicate the effort three times for everything. Thus it makes sense to combine the joint efforts of all
+three teams of kit developers into one. This is the goal of `j5`. `j5` is a single library that provides a uniform interface
+and API to students for all three kits. Whilst code will not be directly portable between the kits, it will also not be
+very hard to port code between them. As a library, `j5` still allows the development teams at the individual competitions
```
Comment (mildlyincompetent) [inline docs/concepts/philosophy.rst]: this sentence successfully says nothing.

File: docs/concepts/philosophy.rst [hunk 5]
```diff
@@ -1,2 +1,40 @@
...
+to have some degree of customisation over how their kit is used.
```
Comment (mildlyincompetent) [inline docs/concepts/philosophy.rst]: On second thoughts I'm okay with the new lines giving my mail client kidney stones, as it enables me to increase the density of my criticism.

File: docs/concepts/philosophy.rst [hunk 6]
```diff
@@ -1,2 +1,40 @@
...
+
+Goals
+-----
+
+There are some goals behind the `j5` project:
+
+- To be compatible with all current and future robotics kits.
```
Comment (mildlyincompetent) [inline docs/concepts/philosophy.rst]: Help help my QuantumBot from the year Three Trillion is incompatible please fix.

File: docs/concepts/philosophy.rst [hunk 7]
```diff
@@ -1,2 +1,40 @@
...
+- To use the latest versions of software and be continuously maintained, even between and during competitions.
```
Comment (mildlyincompetent) [inline docs/concepts/philosophy.rst]: Latest even when unstable? Maybe say "recent" so we don't arch linux ourselves into the depths of hell.

File: docs/concepts/philosophy.rst [hunk 8]
```diff
@@ -1,2 +1,40 @@
...
+- To be an example to students of what good code should look like.
+- To unify all existing robotics kits and simulators into one codebase.
```
Comment (mildlyincompetent) [inline docs/concepts/philosophy.rst]: Does it work with my Dirt Management Robot Simulator from https://endlessrobotics.com/ ?

File: docs/concepts/philosophy.rst (hunk 2 above)
Comment (trickeydan) [reply docs/concepts/philosophy.rst]: I'm disregarding this.

File: docs/concepts/philosophy.rst (hunk 3 above)
Comment (trickeydan) [reply docs/concepts/philosophy.rst]: The robotics competition scene?

File: docs/concepts/philosophy.rst (hunk 4 above)
Comment (trickeydan) [reply docs/concepts/philosophy.rst]: I disagree. The code is portable, just not compatible with all kits due to impossibleness of that.

File: docs/concepts/philosophy.rst (hunk 7 above)
Comment (trickeydan) [reply docs/concepts/philosophy.rst]: Ok, I agree with you here.

File: docs/concepts/philosophy.rst (hunk 8 above)
Comment (trickeydan) [reply docs/concepts/philosophy.rst]: Yes.

File: docs/concepts/philosophy.rst [hunk 9]
```diff
@@ -1,2 +1,25 @@
...
+-----------
+
+Following the reappearance of Student Robotics to the scene in late 2018, there were now three separate, very similar, and also incompatible robotics kits that were being used for the same purpose. None of the kits were perfect, and volunteers didn't want to replicate the effort three times for everything. Thus it makes sense to combine the joint efforts of all three teams of kit developers into one. This is the goal of `j5`. `j5` is a single library that provides a uniform interface and API to students for all three kits. Whilst code will not be directly portable between the kits, it will also not be very hard to port code between them. As a library, `j5` still allows the development teams at the individual competitions to have some degree of customisation over how their kit is used.
+
+Goals
+-----
+
+There are some goals behind the `j5` project:
+
+- To be compatible with all current and future robotics kits.
```
Comment (mildlyincompetent) [inline docs/concepts/philosophy.rst]:
```suggestion
- To be compatible with a variety of relevant current and future robotics kits.
```

**Pull request merged**